### Read data

In [1]:
from core import get_data

# use prepared function for downloading and unzipping the data
url = 'https://storage.googleapis.com/kaggle-competitions-data/kaggle-v2/5407/868283/bundle/archive.zip?GoogleAccessId=web-data@kaggle-161607.iam.gserviceaccount.com&Expires=1583963097&Signature=V1VsrqZZfQ%2Bg0PxuK5VvKo5498kCAz2rlvO9f3fHWa6ib6N6ot%2BdGqasGLwP69wKrhsbgcTQKX13BueMsa6id8EClhBVn7mrdB4GOOLnm0iV1U5gVXDiUNYbvbAhu3dGQY3Yto2a4oNpyJ9EhYOqtGFvIWGDSyhrFm%2Bf3BtySE%2FQs7EwZgJp0S3aMHCoCDcMlwXDvBrQh7rNQo9Ms%2FQ8iOaTUUL1DKKbnfAuyWnTaMQchpoD5P78IRt%2B2uIAQPYfpQx3l4W4lEH5Oo3yZNsQIBfeSBpmsgM5erFl8TRQVrPrhW5pcTjVTNCm8NWSveju2QbgWf8GcfvM6VEmpxzu8g%3D%3D&response-content-disposition=attachment%3B+filename%3Dhouse-prices-advanced-regression-techniques.zip'
get_data(url)

In [2]:
import pandas as pd

house_prices = pd.read_csv('dataset/train.csv')

### Exploratory data analysis

In [3]:
house_prices.shape

(1460, 81)

Dataset has 81 columns and 1460 rows. There is no much data - I'll use cross-validation here

In [4]:
house_prices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

some columns have a lot of missing values

In [5]:
house_prices.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

columns have very different value ranges - I'll use standarization here

In [6]:
house_prices.corr()['SalePrice'].sort_values(ascending=False)

SalePrice        1.000000
OverallQual      0.790982
GrLivArea        0.708624
GarageCars       0.640409
GarageArea       0.623431
TotalBsmtSF      0.613581
1stFlrSF         0.605852
FullBath         0.560664
TotRmsAbvGrd     0.533723
YearBuilt        0.522897
YearRemodAdd     0.507101
GarageYrBlt      0.486362
MasVnrArea       0.477493
Fireplaces       0.466929
BsmtFinSF1       0.386420
LotFrontage      0.351799
WoodDeckSF       0.324413
2ndFlrSF         0.319334
OpenPorchSF      0.315856
HalfBath         0.284108
LotArea          0.263843
BsmtFullBath     0.227122
BsmtUnfSF        0.214479
BedroomAbvGr     0.168213
ScreenPorch      0.111447
PoolArea         0.092404
MoSold           0.046432
3SsnPorch        0.044584
BsmtFinSF2      -0.011378
BsmtHalfBath    -0.016844
MiscVal         -0.021190
Id              -0.021917
LowQualFinSF    -0.025606
YrSold          -0.028923
OverallCond     -0.077856
MSSubClass      -0.084284
EnclosedPorch   -0.128578
KitchenAbvGr    -0.135907
Name: SalePr

'OverallQual' column is very correlated to SalePrice, so I'll split to train / test set by stratifying on this column

In [7]:
from sklearn.model_selection import StratifiedShuffleSplit

split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=93)
for train_index, test_index in split.split(house_prices, house_prices['OverallQual']):
    strat_train_set = house_prices.loc[train_index]
    strat_test_set = house_prices.loc[test_index]

In [8]:
# for now let's rush with some simple model based only on this feature
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score

X_train, y_train = strat_train_set[['OverallQual']], strat_train_set['SalePrice']

lin_reg = LinearRegression(-1)
lin_reg.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=-1, n_jobs=None, normalize=False)

In [9]:
cross_val_score(lin_reg, X_train, y_train, cv=5, scoring='neg_mean_squared_error')

array([-1.80714061e+09, -2.59609559e+09, -2.79391129e+09, -2.01536888e+09,
       -2.13553820e+09])

In [10]:
cross_val_score(lin_reg, X_train, y_train, cv=5, scoring='neg_root_mean_squared_error')

array([-42510.47646066, -50951.89481281, -52857.46202736, -44892.85998838,
       -46211.88377105])

#### Make submission

In [11]:
house_sub = pd.read_csv('dataset/test.csv')

X_sub = house_sub[['OverallQual']]
sub_predictions = lin_reg.predict(X_sub)

In [12]:
result_df = pd.DataFrame({'Id': house_sub['Id'], 'SalePrice': sub_predictions})

In [13]:
result_df.to_csv('first_submission.csv', index=False)